polkadot/test-parachains/adder/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Basic parachain that adds a number as part of its state.
+
+#![no_std]
+
+extern crate polkadot_parachain as parachain;
+extern crate tiny_keccak;
+
+use parachain::codec::{Decode, Encode, Input, Output};
+
+/// Head data for this parachain.
+#[derive(Default, Clone, Hash, Eq, PartialEq)]
+pub struct HeadData {
+	/// Block number
+	pub number: u64,
+	/// parent block keccak256
+	pub parent_hash: [u8; 32],
+	/// hash of post-execution state.
+	pub post_state: [u8; 32],
+}
+
+impl HeadData {
+	pub fn hash(&self) -> [u8; 32] {
+		::tiny_keccak::keccak256(&self.encode())
+	}
+}
+
+impl Encode for HeadData {
+	fn encode_to<T: Output>(&self, dest: &mut T) {
+		dest.push(&self.number);
+		dest.push(&self.parent_hash);
+		dest.push(&self.post_state);
+	}
+}
+
+impl Decode for HeadData {
+	fn decode<I: Input>(input: &mut I) -> Option<Self> {
+		Some(HeadData {
+			number: Decode::decode(input)?,
+			parent_hash: Decode::decode(input)?,
+			post_state: Decode::decode(input)?,
+		})
+	}
+}
+
+/// Block data for this parachain.
+#[derive(Default, Clone)]
+pub struct BlockData {
+	/// State to begin from.
+	pub state: u64,
+	/// Amount to add (overflowing)
+	pub add: u64,
```

Comment:
Why it's overflowing?

Reply:
just to specify!

Reply:
I mean what makes it overflowing? Like, we don't do anything like summing with it

Reply:
Oh, did you mean that it is `%`-ed?

Reply:
it is meant to be an overflowing add

Reply:
ah yeah now it makes sense